python/lsst/daf/butler/gen2convert/walker.py

```diff
@@ -0,0 +1,329 @@
+
+        If the path appears to be a Gen2 Data Repository, a Gen2Data instance
+        representing it will be added to `self.found` (unless it is already there)
+        and returned.  `tryRoot()` also be called recursively on all parent
+        repositories, with all processed repositories added `self.found` and
+        their `.parents` and `.MapperClass` attributes populated.
+
+        Returns `None` if the path does not appear to be the root of a Gen2 Data
+        Repository.
+
+        Returns `True` if the path does appear to be a Gen2 Data Repository but
+        it is in `self.ignored`.
+        """
+        log = Log.getLogger("lsst.daf.butler.gen2convert")
+        root = os.path.abspath(root)
+        repo = self.found.get(root, None)
+        if repo is not None:
+            return repo
+        if root in self.ignored:
+            return True
+
+        MapperClass = None
+        parentPaths = []
+        repoCfgPath = os.path.join(root, "repositoryCfg.yaml")
+        if os.path.exists(repoCfgPath):
+            with open(repoCfgPath, 'r') as f:
+                repoCfg = yaml.load(f)
+            parentPaths = [parent.root for parent in repoCfg.parents]
+            MapperClass = repoCfg.mapper
+        else:
+            parentLinkPath = os.path.join(root, "_parent")
+            if os.path.exists(parentLinkPath):
+                parentPaths.append(os.readlink(parentLinkPath))
+            mapperFilePath = os.path.join(root, "_mapper")
+            if os.path.exists(mapperFilePath):
+                with open(mapperFilePath, 'r') as f:
+                    terms = f.read().strip().split(".")
+                    module = importlib.import_module(".".join(terms[:-1]))
+                    MapperClass = getattr(module, terms[-1])
+
+        # This directory has no _mapper, no _parent, and no repositoryCfg.yaml.
+        # It probably just isn't a repository root.
+        if not (parentPaths or MapperClass):
+            log.debug("%s: not a data repository.", root)
+            return None
+
+        # We now have either a MapperClass or a non-zero parents list, so
+        # it's likely this is a valid repository.  We'll add it to self.found
+        # before we recurse into parents to make sure we short-cut any cycles
+        # (which shouldn't exist, but better to detect that later rather than
+        # recurse infinitely).
+        repo = Gen2Repo(root, MapperClass)
+        self.found[root] = repo
+
+        try:
+            # Recursively construct (or look up) Gen2Repo objects for parents.
+            for path in parentPaths:
+                parent = self.tryRoot(path)
+                if parent is None:
+                    raise ValueError("%s is a parent of %s, but is not a repository." % (parent, root))
+                repo.parents.append(parent)
+            # Make sure there are no cycles in the graph of parents.
+            repo._findCycles()
```

Comment:
Because `tryRoot` short-circuits if `repo` is already in `self.found`, and we add the current `repo` to `self.found` before recursing, it won't find cycles.